packages/jellyfish-transaction/src/script/dftx/dftx_icxorderbook.ts

```diff
@@ -0,0 +1,50 @@
+import { BufferComposer, ComposableBuffer } from '../../buffer/buffer_composer'
+import BigNumber from 'bignumber.js'
+import { Script } from '../../tx'
+import { CScript } from '../../tx_composer'
+
+// Disabling no-return-assign makes the code cleaner with the setter and getter */
+/* eslint-disable no-return-assign */
+
+export enum ICXOrderType {
+  /** type for DFI/BTC orders */
+  INTERNAL = 0x1,
+  /** type for BTC/DFI orders */
+  EXTERNAL = 0x2
+}
+
+/**
+ * ICX CreateOrder DeFi Transaction
+ */
+
+export interface ICXCreateOrder {
+  orderType: number // -------------| 1 byte unsigned, 0x1 (INTERNAL) | 0x2 (EXTERNAL)
+  tokenId: number // ---------------| VarUInt{1-9 bytes}
+  ownerAddress: Script // ----------| n = VarUInt{1-9 bytes}, + n bytes
+  receivePubkey?: string // --------| n = VarUInt{1-9 bytes}, 0x00 (default when undefined) | 0x21 (for COMPRESSED_PUBLIC_KEY_SIZE) | 0x41 (for PUBLIC_KEY_SIZE) + n bytes. See implementation at https://github.com/DeFiCh/ain/blob/ff53dcee23db2ffe0da9b147a0a53956f4e7ee31/src/pubkey.h#L57
+  amountFrom: BigNumber // ---------| 8 bytes unsigned
+  orderPrice: BigNumber // ---------| 8 bytes unsigned
+  expiry: number // ----------------| 4 bytes unsigned
+}
+
+/**
+ * Composable ICXCreateOrder, C stands for Composable.
+ * Immutable by design, bi-directional fromBuffer, toBuffer deep composer.
+ */
+export class CICXCreateOrder extends ComposableBuffer<ICXCreateOrder> {
+  static OP_CODE = 0x31 // '1'
+  static OP_NAME = 'OP_DEFI_TX_ICX_CREATE_ORDER'
+
+  composers (cco: ICXCreateOrder): BufferComposer[] {
+    return [
+      ComposableBuffer.uInt8(() => cco.orderType, v => cco.orderType = v),
+      ComposableBuffer.varUInt(() => cco.tokenId, v => cco.tokenId = v),
+      ComposableBuffer.single<Script>(() => cco.ownerAddress, v => cco.ownerAddress = v, v => new CScript(v)),
+      ComposableBuffer.optionalVarUIntHex(() => cco.receivePubkey, v => cco.receivePubkey = v),
+      ComposableBuffer.satoshiAsBigNumber(() => cco.amountFrom, v => cco.amountFrom = v),
+      ComposableBuffer.satoshiAsBigNumber(() => cco.amountFrom, v => cco.amountFrom = v), // Represents amountToFill: how much is left to fill the order. In case of createOrder, amountToFill is always equal to amountFrom. See https://github.com/DeFiCh/ain/blob/ff53dcee23db2ffe0da9b147a0a53956f4e7ee31/src/masternodes/icxorder.h#L32
```

Comment:
Enforce this on transaction builder (by throwing an error when !=) instead of magically handling it on `CICXCreateOrder`. 

I think a serializer and deserializer should compose and decompose a buffer as it is without any logic. The above interface (`export interface ICXCreateOrder {`) should represent the expected buffer outcome without second guessing. This kind of handling creates intractable buffer management from a evolving code stance.

Reply:
or else simply we forget about the `icx_createorder ` interface.

Reply:
Yes I handled it the same way as C++ side but I can add it to the interface and assert that `amountToFill` === `amountFrom`

Reply:
> or else simply we forget about the `icx_createorder ` interface.

What do you mean by that ?

Reply:
@Jouzo check the `icx_createorder ` rpc, it only has `amountFrom` field.

Reply:
so when crating a new order, `amountToFill ` is redundant

Reply:
Yes exactly, I did it this way to match the RPC interface. 
we can ignore it, add amountToFill to composer interface and assert that amountToFill equals amountFrom

Reply:
I spend a few hours thinking about it, here is my train of thought. 

@surangap pointed out to me that on RPC, it's redundant. However, on DfTx, it's magically pushing redundant buffer into the chain. This was done because it was indexed as such on the storage layer. When a partially offer DfTx is created, the same DfTx is then updated with that partial amountToFill. I guess the same data structure is reused for a different context.

Honestly, not a big fan of this design on chain.

Personally, I rather have a separate indexing serialization layer and transaction sending serialization layer. This implementation creates a debate within me on how `jellyfish-indexer` (2022 Q1 goal) should be implemented whether we should share Transaction DfTx structure with data being indexed. I am standing on the side where we should separate.

However, back to the subject of DfTx. I believe we should still serialize and deserialize accordingly to the interface. The interface should present the buffer explicitly. Double writing and double reading, although make sense, creates unnecessary cognitive complexity and side-effect (think if something was reacting to reading or writing of buffers).

Reply:
Regarding the DfTx, I agree that the serialisation should match the interface.
From the discussion we had, it was not clear for the reviewers where this `amountToFill` was coming from and it is much better to explicitly pass it and check that it is equal to `amountFrom`.
Ideally, the design on chain should be improved as well.

On the indexer matter, I don't grasp the full implication. What this separation would imply ?

Reply:
This is mainly towards jellyfish ecosystem play, as this PR was not fundamentally wrong in any way, it just requires some thinking on my end for aligning goals on how we should implement it #580. Merging this has a profound effect on data indexing.

---

> Providing additional context to the discussion.

As part of the transition https://github.com/DeFiCh/jellyfish/issues/580 (2022 Q4-Q1 long-term planning) where we consolidate whale/ocean/playground into a singular codebase. This transition moves the indexing responsibility to jellyfish. DeFiCh/ain indexes for consensus while jellyfish indexes for consumer/product facing use cases. For consensus, think UTXO unspent set for validating unspent. For product, think querying transaction activity tied to an address. 

Naturally for fast lookup, we have an LSM DB for sorted key querying. 

How it will looked like indexed:
- `[[Key: Data],[Key: Data],[Key: Data]]`
- `[[TxId: Masternode],[TxId: Masternode],[TxId: Masternode]]`

Which is essentially a DfTx:
- `[[TxId: CreateMasternodeDfTx],[TxId: CreateMasternodeDfTx],[TxId: CreateMasternodeDfTx]]`
- `[[TxId: CreatePoolPair],[TxId: CreatePoolPair],[TxId: CreatePoolPair]]`

On the `DeFiCh/whale` side, the indexes were designed rather quickly as we wanted to push out light wallet ASAP. (Build me a light wallet NOW!)
- Redundant double indexing on the whale project, bad
- Use of `JSON.stringify()` for data at REST, boo 👎, waste space
- Requires a high level of heuristic and implementation know-how to index and invalidate
- This leads to the creation of these tickets
  - https://github.com/DeFiCh/whale/issues/50
  - https://github.com/DeFiCh/whale/issues/51

The problems surfaced as fast as they could when Webster, Jimmy, and Ahmed contributed to the code base as development was slow and unnecessarily complex. Take a look at these and you could see how badly the code was, https://github.com/DeFiCh/whale/blob/main/src/module.indexer/model/dftx/update.oracle.ts it works but it sucks.

Pivoting to https://github.com/DeFiCh/jellyfish/issues/580, I did some thinking and I think we should fall back to indexing as similar as possible to how DeFiCh/ain does it. Look above, via `[Key: Data]`. **Where we use BufferComposer and store buffer as data in the indexes.**

Jumping to the subject on separation.

`DeFiCh/ain` indexes consensus protocol heavy object. Consensus protocol heavy objects are pretty straightforward, you index something with consistent lookup behavior. Think `txid`, always consistent E.g. `[TxId: DfTx]`: `[TxId: ICXCreateOrder]` 

Given you need to maintain a list of `DfTx` e.g. `ICX Order List`
Having the same DfTx, the structure allows a list of data as such: `[[TxId: ICXOrder], [TxId: ICXOrder]]`

Using the same DfTx abstraction: 
- `ICXCreateOrder extends ICXOrder` creates a new ICXOrder and add into `[[TxId: ICXOrder], [TxId: ICXOrder]]`
- `ICXUpdateOrder extends ICXOrder` with txId updates the same ICXOrder in `[[TxId: ICXOrder], [TxId: ICXOrder]]` via lookup.

`TxId` being as good as a random id, makes querying in a sorted list rather useless from the product API perspective. Meanless order.

This PR creates an inflection towards #580 (playing devil advocate with myself using newfound information from this PR), should we use the same `DfTx` for on-chain broadcasting and indexing or create another buffer interface purely for indexing. Initially, I wanted to just store `[MeaningfulKey: DfTx]`. This PR just signals the need for separation to design better consumer-friendly product APIs.

Reply:
@fuxingloh how does this impact indexers that don't rely directly on tx data, e.g. Aggregated Oracle Prices, Masternode stats, etc. Would those still live on jellyfish?

Reply:
Yup, leading up to that path, they will be on jellyfish too. The separations are inevitable where product-facing APIs need to index non-Tx-based data.